src/zkp/schnorr_adaptor.rs

```diff
@@ -0,0 +1,298 @@
+//! Bindings for Schnorr based adaptor signatures in secp256k1-zkp.
+
+use crate::ffi::{self, CPtr, SCHNORR_ADAPTOR_PRESIGNATURE_LENGTH};
+#[cfg(feature = "rand-std")]
+use crate::rand::thread_rng;
+#[cfg(feature = "actual-rand")]
+use crate::rand::{CryptoRng, Rng};
+use crate::{PublicKey, Secp256k1, SecretKey, Keypair, XOnlyPublicKey};
+use crate::constants::{SECRET_KEY_SIZE,SCHNORR_SIGNATURE_SIZE};
+use crate::schnorr::Signature as SchnorrSignature;
+use crate::{Message, Signing};
+use crate::{from_hex, Error};
+use core::{fmt, ptr, str};
+
+/// Represents an adaptor signature
+#[derive(Debug, PartialEq, Clone, Copy, Eq)]
+pub struct SchnorrAdaptorPreSignature(ffi::SchnorrAdaptorPreSignature);
+
+impl fmt::LowerHex for SchnorrAdaptorPreSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for ch in self.0.as_ref().iter() {
+            write!(f, "{:02x}", ch)?;
+        }
+        Ok(())
+    }
+}
+
+impl fmt::Display for SchnorrAdaptorPreSignature {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl str::FromStr for SchnorrAdaptorPreSignature {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<SchnorrAdaptorPreSignature, Error> {
```

Comment:
In 28fd225cc653682abc5b320f7c45c0ca2d0b7b36:

Can you use `Self` rather than `SchnorrAdaptorPreSignature` everywhere you can? It helps to keep track of which functions are manipulating an object and which ones are converting it to other types.